core/trino-main/src/test/java/io/trino/execution/scheduler/TestFaultTolerantStageScheduler.java

```diff
@@ -522,6 +541,229 @@ public void testReportTaskFailure()
+            assertBlocked(blocked);
+            assertThat(remoteTaskFactory.getTasks()).hasSize(6);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 0)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 1)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 2)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 3)).getTaskStatus().getState(), TaskState.RUNNING);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(1, 0)).getTaskStatus().getState(), TaskState.RUNNING);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(2, 0)).getTaskStatus().getState(), TaskState.RUNNING);
+
+            // T+6.6 task 1 failure
+            scheduler.reportTaskFailure(getTaskId(1, 0), new RuntimeException("some other failure"));
+            assertUnblocked(blocked);
+            scheduler.schedule();
+            blocked = scheduler.isBlocked();
+            assertBlocked(blocked);
+            assertThat(remoteTaskFactory.getTasks()).hasSize(6);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 0)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 1)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 2)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 3)).getTaskStatus().getState(), TaskState.RUNNING);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(1, 0)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(2, 0)).getTaskStatus().getState(), TaskState.RUNNING);
+
+            // T+9.0 task1 still not retried (delay is 3s)
+            moveTime(2400, MILLISECONDS);
+            assertBlocked(blocked);
+            assertThat(remoteTaskFactory.getTasks()).hasSize(6);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 0)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 1)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 2)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(0, 3)).getTaskStatus().getState(), TaskState.RUNNING);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(1, 0)).getTaskStatus().getState(), TaskState.FAILED);
+            assertEquals(remoteTaskFactory.getTasks().get(getTaskId(2, 0)).getTaskStatus().getState(), TaskState.RUNNING);
+
+            // T+9.0 task 0.3 completes successfully - should kill delay for task 1
```

Comment:
For my understanding, the delay is totally reset for the whole stage, right?

Reply:
yep - the comment here is misleading - will make it clear.